### ISSUE TITLE
Unable to achieve consensus because of state queue

### ISSUE THREAD

**Issue by viquezclaudio** [nimiq/core-rs-albatross#2504]
In the current testnet,  full nodes are unable to achieve consensus because the state sync is never completed (it never starts)

The following behavior is observed in the attached logs:

- Validity syncing is complete
- Nodes are moved to live sync
- We continue pushing blocks until we catch the tip of the chain
- Consensus is never achieved.

Also, it is important to note that the following log messages are **not** observed:
`Trie incomplete, starting state sync.`
`Finished state sync, trie complete.`

As part of the debug process, I added extra log messages to the state queue poll_next function to dump the state of the _start_key.is_complete()_ and _blockchain_state_complete_ variables.
Both remain false during the whole time, and their state is never observed to change, which explains why those log messages are never observed as per this [logic](https://github.com/nimiq/core-rs-albatross/blob/albatross/consensus/src/sync/live/state_queue/mod.rs#L492)

Some `Discarding chunk, block already applied` messages are also, occasionally, observed. 

[no-consensus-state-queue.log](https://github.com/nimiq/core-rs-albatross/files/15395668/no-consensus-state-queue.log)

**Comment by viquezclaudio** [comment]
After adding more logs to the state queue it is observed that chunks are continuously being requested:
```
Going to request a chunk, start_key Some(0e9f95)
Going to request a chunk, start_key Some(0ea19)
 Going to request a chunk, start_key Some(0ea396)
```

However, there's still no observed progress in the state queue

**Comment by viquezclaudio** [comment]
After further debugging and adding more logs, it can be observed that chunks are being pushed into the state queue buffer

`21:11:50 DEBUG state_queue          | Going to request a chunk, start_key Some(0731147)
21:11:50 DEBUG chunk_request_compo… |  Requesting mising chunks from peer: 12D3KooWReHe8Yb6zTEyrbA6KLAvu1KhG1znZnz8cm5FiW2GJY5i
21:11:51 DEBUG block_queue          | Received block via gossipsub block=#5028696:MI:3443729648 peer_id=12D3KooWKd6cY4saxrTradfE5ELkxDFNWTC3Vn79H8ckNQDREA75
21:11:51 DEBUG chunk_request_compo… | Polling sync queue... 
21:11:51 DEBUG chunk_request_compo… | obtained response from sync queue 
21:11:51 DEBUG state_queue          |  Obtained poll ready for chunk from 12D3KooWReHe8Yb6zTEyrbA6KLAvu1KhG1znZnz8cm5FiW2GJY5i, start key0731147
21:11:51 DEBUG state_queue          |  We recieved a response chunk
21:11:51 DEBUG state_queue          |  We have a good chunk Chunk { block_number: 5028696, block_hash: 3443729648d96ca484d86f2e824856ac2d1fe426c3ebcc04bce38f3339a5a30b, chunk: TrieChunk { end_key: Some(0732f), items: #200, proof: .. } }
21:11:51 DEBUG state_queue          | Going to insert chunk into the buffer
21:11:51 DEBUG state_queue          | Inserted chunk TrieChunk { end_key: Some(0732f), items: #200, proof: .. } into the buffer, new buffer size 1
21:11:51 DEBUG state_queue          | Going to request a chunk, start_key Some(0732f)
21:11:51 DEBUG chunk_request_compo… |  Requesting mising chunks from peer: 12D3KooWEjugkhfhFs9RyWvi2zdWnDEJgjHdgNsYRXHm7EVt8jk1
21:11:51 DEBUG chunk_request_compo… | Polling sync queue... 
21:11:51 DEBUG chunk_request_compo… | obtained response from sync queue 
21:11:51 DEBUG state_queue          |  Obtained poll ready for chunk from 12D3KooWEjugkhfhFs9RyWvi2zdWnDEJgjHdgNsYRXHm7EVt8jk1, start key0732f
21:11:51 DEBUG state_queue          |  We recieved a response chunk
21:11:51 DEBUG state_queue          |  We have a good chunk Chunk { block_number: 5028696, block_hash: 3443729648d96ca484d86f2e824856ac2d1fe426c3ebcc04bce38f3339a5a30b, chunk: TrieChunk { end_key: Some(0734fa), items: #200, proof: .. } }
21:11:51 DEBUG state_queue          | Going to insert chunk into the buffer
21:11:51 DEBUG state_queue          | Inserted chunk TrieChunk { end_key: Some(0734fa), items: #200, proof: .. } into the buffer, new buffer size 2
21:11:51 DEBUG state_queue          | Going to request a chunk, start_key Some(0734fa)
21:11:51 DEBUG chunk_request_compo… |  Requesting mising chunks from peer: 12D3KooWLa1EjUs1dV7bXW39X6oar8QtqiXX6L87vCZ2cxX9A9Vo
21:11:51 DEBUG push                 | Accepted block block=#5028696:MI:3443729648 num_transactions=2 kind="extend"
21:11:51 INFO  state_queue          |  State queue poll function blockchain_state_complete=false start_key_complete=false
21:11:51 DEBUG block_queue          | Broadcasting on BlockHeaderTopic block="#5028696:MI:3443729648"
21:11:51 DEBUG block_queue          | Failed to publish block header block="#5028696:MI:3443729648" error=Gossipsub Publish error: InsufficientPeers
21:11:51 DEBUG block_queue          | Received block via gossipsub block=#5028697:MI:1d4b8c71ad peer_id=12D3KooWReHe8Yb6zTEyrbA6KLAvu1KhG1znZnz8cm5FiW2GJY5i
21:11:51 DEBUG chunk_request_compo… | Polling sync queue... 
21:11:51 DEBUG chunk_request_compo… | obtained response from sync queue 
21:11:51 DEBUG state_queue          |  Obtained poll ready for chunk from 12D3KooWLa1EjUs1dV7bXW39X6oar8QtqiXX6L87vCZ2cxX9A9Vo, start key0734fa
21:11:51 DEBUG state_queue          |  We recieved a response chunk
21:11:51 DEBUG state_queue          |  We have a good chunk Chunk { block_number: 5028697, block_hash: 1d4b8c71adad6cc6b4962c7900fe6f448b4ad8d75a5c7c4b66e5c7e6eb182792, chunk: TrieChunk { end_key: Some(0736c8), items: #200, proof: .. } }
21:11:51 DEBUG state_queue          | Going to insert chunk into the buffer
21:11:51 DEBUG state_queue          | Inserted chunk TrieChunk { end_key: Some(0736c8), items: #200, proof: .. } into the buffer, new buffer size 1
21:11:51 DEBUG state_queue          | Going to request a chunk, start_key Some(0736c8)
21:11:51 DEBUG chunk_request_compo… |  Requesting mising chunks from peer: 12D3KooWF4pZSd4wPcTqwhWVCBkV3T4qWdzdQSKpyHE342c64xeA
21:11:52 DEBUG push                 | Accepted block block=#5028697:MI:1d4b8c71ad num_transactions=35 kind="extend"
21:11:52 INFO  state_queue          |  State queue poll function blockchain_state_complete=false start_key_complete=false
21:11:52 DEBUG block_queue          | Broadcasting on BlockHeaderTopic block="#5028697:MI:1d4b8c71ad"
21:11:52 DEBUG block_queue          | Failed to publish block header block="#5028697:MI:1d4b8c71ad" error=Gossipsub Publish error: InsufficientPeers
21:11:52 INFO  nimiq_client         | Consensus: lost - Head: #5028697:MI:1d4b8c71ad - Peers: 13 consensus_established=false block_number=5028697 num_peers=13
21:11:52 DEBUG chunk_request_compo… | Polling sync queue... 
21:11:52 DEBUG chunk_request_compo… | obtained response from sync queue 
21:11:52 DEBUG state_queue          |  Obtained poll ready for chunk from 12D3KooWF4pZSd4wPcTqwhWVCBkV3T4qWdzdQSKpyHE342c64xeA, start key0736c8
21:11:52 DEBUG state_queue          |  We recieved a response chunk
21:11:52 DEBUG state_queue          |  We have a good chunk Chunk { block_number: 5028697, block_hash: 1d4b8c71adad6cc6b4962c7900fe6f448b4ad8d75a5c7c4b66e5c7e6eb182792, chunk: TrieChunk { end_key: Some(07389f), items: #200, proof: .. } }`

**Comment by viquezclaudio** [comment]
After adding more logs, it is observed that chunks are being received and pushed into the state queue buffer:


`
21:11:50 state_queue                      | Going to request a chunk, start_key Some(0731147)
21:11:50 chunk_request_compo… |  Requesting mising chunks from peer: 12D3KooWReHe8Yb6zTEyrbA6KLAvu1KhG1znZnz8cm5FiW2GJY5i
21:11:51 block_queue                      | Received block via gossipsub block=#5028696:MI:3443729648 peer_id=12D3KooWKd6cY4saxrTradfE5ELkxDFNWTC3Vn79H8ckNQDREA75
21:11:51 chunk_request_compo…  | Polling sync queue... 
21:11:51 chunk_request_compo…  | obtained response from sync queue 
21:11:51 state_queue                       |  Obtained poll ready for chunk from 12D3KooWReHe8Yb6zTEyrbA6KLAvu1KhG1znZnz8cm5FiW2GJY5i, start key0731147
21:11:51 state_queue                       |  We recieved a response chunk
21:11:51 state_queue                       |  We have a good chunk Chunk { block_number: 5028696, block_hash: 3443729648d96ca484d86f2e824856ac2d1fe426c3ebcc04bce38f3339a5a30b, chunk: TrieChunk { end_key: Some(0732f), items: #200, proof: .. } }
21:11:51 state_queue                       | Going to insert chunk into the buffer
21:11:51 state_queue                       | Inserted chunk TrieChunk { end_key: Some(0732f), items: #200, proof: .. } into the buffer, new buffer size 1
21:11:51 state_queue                       | Going to request a chunk, start_key Some(0732f)
21:11:51 chunk_request_compo…  |  Requesting mising chunks from peer: 12D3KooWEjugkhfhFs9RyWvi2zdWnDEJgjHdgNsYRXHm7EVt8jk1
21:11:51 chunk_request_compo…  | Polling sync queue... 
21:11:51 chunk_request_compo…  | obtained response from sync queue 
21:11:51 state_queue                       |  Obtained poll ready for chunk from 12D3KooWEjugkhfhFs9RyWvi2zdWnDEJgjHdgNsYRXHm7EVt8jk1, start key0732f
21:11:51 state_queue                       |  We recieved a response chunk
21:11:51 state_queue                       |  We have a good chunk Chunk { block_number: 5028696, block_hash: 3443729648d96ca484d86f2e824856ac2d1fe426c3ebcc04bce38f3339a5a30b, chunk: TrieChunk { end_key: Some(0734fa), items: #200, proof: .. } }
21:11:51 state_queue                       | Going to insert chunk into the buffer
21:11:51 state_queue                       | Inserted chunk TrieChunk { end_key: Some(0734fa), items: #200, proof: .. } into the buffer, new buffer size 2
21:11:51 state_queue                       | Going to request a chunk, start_key Some(0734fa)
21:11:51 chunk_request_compo…  |  Requesting mising chunks from peer: 12D3KooWLa1EjUs1dV7bXW39X6oar8QtqiXX6L87vCZ2cxX9A9Vo
21:11:51 push                                    | Accepted block block=#5028696:MI:3443729648 num_transactions=2 kind="extend"
21:11:51 state_queue                       |  State queue poll function blockchain_state_complete=false start_key_complete=false
21:11:51 block_queue                      | Broadcasting on BlockHeaderTopic block="#5028696:MI:3443729648"
21:11:51 block_queue                      | Received block via gossipsub block=#5028697:MI:1d4b8c71ad peer_id=12D3KooWReHe8Yb6zTEyrbA6KLAvu1KhG1znZnz8cm5FiW2GJY5i
21:11:51 chunk_request_compo…  | Polling sync queue... 
21:11:51 chunk_request_compo…  | obtained response from sync queue 
21:11:51 state_queue                       |  Obtained poll ready for chunk from 12D3KooWLa1EjUs1dV7bXW39X6oar8QtqiXX6L87vCZ2cxX9A9Vo, start key0734fa
21:11:51 state_queue                       |  We recieved a response chunk
21:11:51 state_queue                       |  We have a good chunk Chunk { block_number: 5028697, block_hash: 1d4b8c71adad6cc6b4962c7900fe6f448b4ad8d75a5c7c4b66e5c7e6eb182792, chunk: TrieChunk { end_key: Some(0736c8), items: #200, proof: .. } }
21:11:51 state_queue                        | Going to insert chunk into the buffer
21:11:51 state_queue                        | Inserted chunk TrieChunk { end_key: Some(0736c8), items: #200, proof: .. } into the buffer, new buffer size 1
21:11:51 state_queue                        | Going to request a chunk, start_key Some(0736c8)
21:11:51 chunk_request_compo…   |  Requesting mising chunks from peer: 12D3KooWF4pZSd4wPcTqwhWVCBkV3T4qWdzdQSKpyHE342c64xeA
21:11:52 push                                    | Accepted block block=#5028697:MI:1d4b8c71ad num_transactions=35 kind="extend"
21:11:52 state_queue                       |  State queue poll function blockchain_state_complete=false start_key_complete=false
21:11:52 block_queue                      | Broadcasting on BlockHeaderTopic block="#5028697:MI:1d4b8c71ad"
21:11:52 nimiq_client                       | Consensus: lost - Head: #5028697:MI:1d4b8c71ad - Peers: 13 consensus_established=false block_number=5028697 num_peers=13


`

**Comment by sisou** [comment]
How is it completed, though?

**Comment by viquezclaudio** [comment]
This is the issue of state sync takes too long... but it completes, after ~5 hours in my case using the testnet 2 weeks ago

**Comment by sisou** [comment]
So the state sync actually runs in the background, and the initial report of this issue was invalid. And it's basically "fixed" by #2532?

**Comment by viquezclaudio** [comment]
Exactly!
